runtime/common/src/impl_multiasset_paymaster.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2025 Moonbeam foundation
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use frame_support::traits::OriginTrait;
+use frame_support::traits::{fungible::NativeOrWithId, tokens::Pay};
+use moonbeam_core_primitives::{AssetId, Balance};
+use pallet_moonbeam_foreign_assets::AssetsById;
+use sp_core::U256;
+use sp_runtime::{traits::StaticLookup, DispatchError};
+
+pub struct MultiAssetPaymaster<R>(sp_std::marker::PhantomData<R>);
+impl<R> Pay for MultiAssetPaymaster<R>
+where
+	R: frame_system::Config
+		+ pallet_treasury::Config
+		+ pallet_balances::Config
+		+ pallet_asset_manager::Config
+		+ pallet_assets::Config
+		+ pallet_moonbeam_foreign_assets::Config
+		+ pallet_xcm_weight_trader::Config,
+{
+	type Balance = Balance;
+	type Beneficiary = R::AccountId;
+	type AssetKind = NativeOrWithId<AssetId>;
+	type Id = ();
+	type Error = DispatchError;
+	fn pay(
+		who: &Self::Beneficiary,
+		asset_kind: Self::AssetKind,
+		amount: Self::Balance,
+	) -> Result<Self::Id, Self::Error> {
+		match asset_kind {
+			Self::AssetKind::Native => {
+				let dest = R::Lookup::unlookup(who.clone());
+				let value = <R as pallet_balances::Config>::Balance::try_from(amount)
+					.map_err(|_| pallet_treasury::Error::<R>::PayoutError)?;
+				// Pay account with native balance
+				pallet_balances::Pallet::<R>::transfer_keep_alive(
+					R::RuntimeOrigin::signed(pallet_treasury::Pallet::<R>::account_id()),
+					dest,
+					value,
+				)?;
+				Ok(())
+			}
+			Self::AssetKind::WithId(id) => {
+				// Check in the foreign assets first
+				if let Some(_asset_loc) = AssetsById::<R>::get(id) {
```

Comment:
I feel that it would be cleaner to have simple traits for foreign asset (or "non-native asset", if we want to avoid the foreign asset terminology in this context) lookup, transfer, creation & mint and pass types implementing these instead of hard-coding the logic and the dependencies to `pallet_moonbeam_foreign_assets` and `pallet_moonbeam_foreign_assets`. wdyt ?